Software/app/basic_fuota/lora_app.c

```diff
@@ -0,0 +1,528 @@
+    .FragDecoderRead = FragDecoderRead,
+  },
+
+  .OnProgress = OnFragProgress,
+  .OnDone = OnFragDone
+};
+
+/*
+ * Indicates if LoRaMacProcess call is pending.
+ *
+ * warning If variable is equal to 0 then the MCU can be set in low power mode
+ */
+static volatile uint8_t IsMacProcessPending = 0;
+
+/*!
+ * Indicates if a Tx frame is pending.
+ *
+ * \warning Set to 1 when OnTxTimerEvent raised
+ */
+static volatile uint8_t IsTxFramePending = 0;
+
+/*
+ * Indicates if the system time has been synchronized
+ */
+static volatile bool IsClockSynched = false;
+
+/*
+ * MC Session Started
+ */
+static volatile bool IsMcSessionStarted = false;
+
+/*
+ * Indicates if the file transfer is done
+ */
+static volatile bool IsFileTransferDone = false;
+
+/*
+ *  Received file computed CRC32
+ */
+static volatile uint32_t FileRxCrc = 0;
+
+/**
+  * @brief Timer to handle the application Tx
+  */
+static UTIL_TIMER_Object_t TxTimer;
+
+void LoRaWAN_Init(void)
+{
+  UTIL_SEQ_RegTask((1 << CFG_SEQ_Task_LmHandlerProcess), UTIL_SEQ_RFU, LmHandlerProcess);
+  UTIL_SEQ_RegTask((1 << CFG_SEQ_Task_LoRaSendOnTxTimer), UTIL_SEQ_RFU, SendTxData);
+
+  /* Init Info table used by LmHandler*/
+  LoraInfo_Init();
+
+  /* Init the Lora Stack*/
+  LmHandlerInit(&LmHandlerCallbacks);
+
+  LmHandlerPackageRegister(PACKAGE_ID_CLOCK_SYNC, NULL);
+
+  LmHandlerPackageRegister(PACKAGE_ID_REMOTE_MCAST_SETUP, NULL);
+
+  LmHandlerPackageRegister(PACKAGE_ID_FRAGMENTATION, &FragmentationParams);
+
+  LmHandlerConfigure(&LmHandlerParams);
+
+  /* state variable to indicate synchronization done*/
+  IsClockSynched = false;
+
+  /* state variable to indicate data block transfer done*/
+  IsFileTransferDone = false;
+
+  LmHandlerJoin(ActivationType);
+
+  /* send every time timer elapses */
+  UTIL_TIMER_Create(&TxTimer, 0xFFFFFFFFU, UTIL_TIMER_ONESHOT, OnTxTimerEvent, NULL);
+  UTIL_TIMER_SetPeriod(&TxTimer, APP_TX_DUTYCYCLE);
+  UTIL_TIMER_Start(&TxTimer);
+}
+
+static void OnRxData(LmHandlerAppData_t *appData, LmHandlerRxParams_t *params)
+{
+  if ((appData != NULL) && (params != NULL))
+  {
+    static const char *slotStrings[] = {"1", "2", "C", "C Multicast", "B Ping-Slot", "B Multicast Ping-Slot"};
+
+    APP_LOG(ADV_TRACER_TS_OFF, ADV_TRACER_VLEVEL_M, "\r\n ###### ========== MCPS-Indication ==========\r\n");
+    APP_LOG(ADV_TRACER_TS_OFF, ADV_TRACER_VLEVEL_M, "\r\n ###### D/L FRAME:%04d | SLOT:%s | PORT:%d | DR:%d | RSSI:%d | SNR:%d\r\n",
+            params->DownlinkCounter, slotStrings[params->RxSlot], appData->Port, params->Datarate, params->Rssi, params->Snr);
+  }
+}
+
+static void SendTxData(void)
+{
+  LmHandlerErrorStatus_t status = LORAMAC_HANDLER_ERROR;
+
+  uint8_t isPending = 0;
+  CRITICAL_SECTION_BEGIN();
+  isPending = IsTxFramePending;
+  IsTxFramePending = 0;
+  CRITICAL_SECTION_END();
+  if (isPending == 1)
+  {
+    if (LmHandlerIsBusy() == true)
+    {
+      return;
+    }
+
+    if (IsMcSessionStarted == false)    /* we are in Class A*/

```

Comment:
Nitpicking: addressing as 'we' is kinda weird (happens a few times, but only in this function). Maybe change this to more formal language

Reply:
Sure